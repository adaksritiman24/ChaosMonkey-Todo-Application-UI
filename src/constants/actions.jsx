export const ACTIONS = {
    CHANGE_SCOPE : "change scope",
    SET_IS_CREATE_TODO_MODAL_OPEN : "set is create todo modal open",
    SET_TODOS : "set todos",
}
