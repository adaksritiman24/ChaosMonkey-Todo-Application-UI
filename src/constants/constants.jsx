export const SCOPE = {
    ALL : "all",
    COMPLETED : "completed",
    INCOMPLETE : "incomplete",
}

export const state = {
    scope : SCOPE.ALL,
    isCreateTodoModalOpen : false,
    loading : true,
}

export const baseURI = "http://localhost:8080/todoservice/v1";