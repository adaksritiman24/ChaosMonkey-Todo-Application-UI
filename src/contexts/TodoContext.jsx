import { createContext } from "react";

const initialContextValue = {
    state : null,
    dispatch : ()=> {}
}

const TodoContext = createContext(initialContextValue);

export default TodoContext;