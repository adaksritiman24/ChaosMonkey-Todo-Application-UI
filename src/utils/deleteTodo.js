import axios from "axios";
import { baseURI } from "../constants/constants"

export const deleteTodo = async(id)=> {
    const config = {
        method : "delete",
        url : baseURI + `/todo/${id}`,
        headers : {} 
    };

    const {data} = await axios(config);
    return data.todoId;
}
