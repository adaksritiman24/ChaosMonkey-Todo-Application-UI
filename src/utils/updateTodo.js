import axios from "axios";
import { baseURI } from "../constants/constants";

const updateTodo= async(newTodo)=>{
    const todoData = JSON.stringify({
        "title": newTodo.title,
        "body": newTodo.body,
        "completed": newTodo.completed,
      });

     const config = {
        method: 'patch',
        url: baseURI+'/todo/'+newTodo.id,
        headers: { 
          'Content-Type': 'application/json'
        },
        data : todoData,
      };

    const {data} =await axios(config);
    return data;
}

export default updateTodo;