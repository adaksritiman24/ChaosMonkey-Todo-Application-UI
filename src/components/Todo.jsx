import { Edit } from "@mui/icons-material";
import {
  Button,
  Card,
  CardActions,
  CardContent,
  Grid,
  IconButton,
  Paper,
  Stack,
  styled,
  Typography,
} from "@mui/material";
import { green, grey } from "@mui/material/colors";
import { Box } from "@mui/system";
import React,{useContext} from "react";
import { ACTIONS } from "../constants/actions";
import TodoContext from "../contexts/TodoContext";
import updateTodo from "../utils/updateTodo";

const StyledPaper = styled(Paper)(({ theme }) => ({
  height: "250px",
  backgroundColor: grey[200],
}));

const buttonStyles = {
  borderRadius : "20px",
  boxShadow : "none",
  textTransform : "none",
}

function Todo({todoDetails}) {

  const {appState, dispatch} = useContext(TodoContext);

  const setTodoStatus = async(completed)=> {
    const updatedTodo = {...todoDetails, completed};
    const updatedTodoResponse = await updateTodo(updatedTodo);

    const updatedTodos = appState.todos.map((todo)=>{
      if(todo.id === updatedTodoResponse.id)
        return {...todo, completed: updatedTodoResponse.completed};
      return todo;
    });

    dispatch({
      type : ACTIONS.SET_TODOS,
      payload : updatedTodos
    })
  }


  return (
    <Grid item lg={4} sm={6} xs={12}>
      <StyledPaper elevation={2}>
        <Card
          sx={{
            height: "100%",
            
          }}
        >
          <CardContent 
          sx={{
            height : "150px",
            maxHeight : "150px",
            overflow : "hidden",
            backgroundColor : todoDetails.completed===1? green[50] : grey[0],
          }}
            
          >
            <Typography gutterBottom variant="h5" component="div"
              sx={{
                textOverflow : "ellipsis",
                WebkitLineClamp: 2,
                overflow: "hidden",
                display: "-webkit-box",
                WebkitBoxOrient: "vertical",

              }}
            >
            {todoDetails.title}
            </Typography>
            <Typography
               variant="body2" 
               color="text.secondary"
               sx={{
                WebkitLineClamp: 4,
                overflow: "hidden",
                display: "-webkit-box",
                WebkitBoxOrient: "vertical",
               }}
            >
            {todoDetails.body}
            </Typography>
          </CardContent>

          <CardActions
            sx={{
              height : "50px",
              display : "flex",
              justifyContent : "space-between",
            }}
          >
            <Stack spacing={1} direction="row">
              <Button 
              color="success" 
              variant={todoDetails.completed===1? "contained" : "outlined"} 
              sx={buttonStyles} 
              onClick={()=>setTodoStatus(1)}>
                Completed
              </Button>
              <Button 
              color="error" 
              variant={todoDetails.completed===0? 
              "contained" : "outlined"} 
              sx={buttonStyles}  
              onClick={()=>setTodoStatus(0)}>
                Incomplete
              </Button>
            </Stack>
            <Box>
              <IconButton>
                <Edit/>
              </IconButton>
            </Box>
          </CardActions>
        </Card>
      </StyledPaper>
    </Grid>
  );
}

export default Todo;
