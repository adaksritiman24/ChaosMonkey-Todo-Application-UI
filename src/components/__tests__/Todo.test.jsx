import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import TodoContext from "../../contexts/TodoContext";
import { SCOPE } from "../../constants/constants";
import Todo from "../Todo";
import { ACTIONS } from "../../constants/actions";
import { act } from "react-dom/test-utils";
import * as UpdateTodoUtils from "../../utils/updateTodo";

const dummyTodos = [
  {
    id: 1,
    title: "Dummy Title",
    body: "Dummy Body",
    completed: 0,
  },
  {
    id: 2,
    title: "Dummy Title 2",
    body: "Dummy Body 2",
    completed: 1,
  },
];

const dummyAppState = {
  todos: dummyTodos,
  scope: SCOPE.ALL,
  isUpdateTodoModalOpen: false,
};
const dispatch = jest.fn();

const TodoContextProvider = (props) => {
  const appState = { ...dummyAppState, ...props.modifiedState };

  const todoContextValue = { appState, dispatch };
  return (
    <TodoContext.Provider value={todoContextValue}>
      {props.children}
    </TodoContext.Provider>
  );
};

const renderTodoWithDummyTodoDetails = () => {
  const dummyTodoDetail = dummyTodos[0];
  const { baseElement } = render(
    <TodoContextProvider>
      <Todo todoDetails={dummyTodoDetail} />
    </TodoContextProvider>
  );

  return baseElement;
};

describe("Todo", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("Should render Todo item when todoDetail is passed as prop", () => {
    const baseElement = renderTodoWithDummyTodoDetails();

    expect(screen.getByText(/Dummy Title/)).toBeInTheDocument();
    expect(screen.getByText(/Dummy Body/)).toBeInTheDocument();

    expect(baseElement).toMatchSnapshot();
  });

  it("Should dispatch action when Edit icon is clicked", () => {
    renderTodoWithDummyTodoDetails();
    const editButton = screen.getAllByRole("button")[2];

    fireEvent.click(editButton);

    expect(dispatch).toBeCalledTimes(2);
    expect(dispatch).toBeCalledWith({
      type: ACTIONS.SET_IS_UPDATE_TODO_MODAL_OPEN,
      payload: true,
    });
    expect(dispatch).toBeCalledWith({
      type: ACTIONS.SET_UPDATING_TODO,
      payload: dummyTodos[0],
    });
  });

  it("Should dispatch updated todolist when complete button is clicked", async () => {
    const updatedTodoDetails = {
      ...dummyTodos[0],
      completed: 1,
    };
    jest
      .spyOn(UpdateTodoUtils, "updateTodo")
      .mockImplementation(() => Promise.resolve(updatedTodoDetails));

    renderTodoWithDummyTodoDetails();

    const expectedListOfTodos = [
      {
        id: 1,
        title: "Dummy Title",
        body: "Dummy Body",
        completed: 1,
      },
      {
        id: 2,
        title: "Dummy Title 2",
        body: "Dummy Body 2",
        completed: 1,
      },
    ];

    const completeButton = screen.getAllByRole("button")[0];
    act(() => {
      fireEvent.click(completeButton);
    });

    await waitFor(() => {
      expect(dispatch).toBeCalledWith({
        type: ACTIONS.SET_TODOS,
        payload: expectedListOfTodos,
      });
    });
  });
});
