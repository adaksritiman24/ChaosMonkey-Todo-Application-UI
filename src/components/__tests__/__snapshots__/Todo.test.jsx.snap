// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo Should render Todo item when todoDetail is passed as prop 1`] = `
<body>
  <div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-lg-4 css-13i4rnv-MuiGrid-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 css-1mzto2f-MuiPaper-root"
        style="opacity: 1; transform: scale(1, 1); transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-zwt9xa-MuiPaper-root-MuiCard-root"
          data-testid="todo-item-card"
        >
          <div
            class="MuiCardContent-root css-xxg32x-MuiCardContent-root"
          >
            <div
              class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-r80n0a-MuiTypography-root"
            >
              Dummy Title
            </div>
            <p
              class="MuiTypography-root MuiTypography-body2 css-3pr4ab-MuiTypography-root"
            >
              Dummy Body
            </p>
          </div>
          <div
            class="MuiCardActions-root MuiCardActions-spacing css-v9gbui-MuiCardActions-root"
          >
            <div
              class="css-w4z10b-MuiStack-root"
            >
              <button
                class="MuiButton-root MuiButton-outlined MuiButton-outlinedSuccess MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-yh0l3o-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                Completed
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
              <button
                class="MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-1amz1tr-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                Incomplete
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <div
              class="MuiBox-root css-0"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="EditIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
